java/ql/src/Security/CWE/CWE-730/PolynomialReDoS.ql

```diff
@@ -0,0 +1,48 @@
+/**
+ * @name Polynomial regular expression used on uncontrolled data
+ * @description A regular expression that can require polynomial time
+ *              to match may be vulnerable to denial-of-service attacks.
+ * @kind path-problem
+ * @problem.severity warning
+ * @precision high
+ * @id java/polynomial-redos
+ * @tags security
+ *       external/cwe/cwe-730
+ *       external/cwe/cwe-400
+ */
+
+import java
+import semmle.code.java.security.performance.SuperlinearBackTracking
+import semmle.code.java.dataflow.DataFlow
+import semmle.code.java.regex.RegexTreeView
+import semmle.code.java.regex.RegexFlowConfigs
+import semmle.code.java.dataflow.FlowSources
+import DataFlow::PathGraph
+
+class PolynomialRedosSink extends DataFlow::Node {
+  RegExpLiteral reg;
+
+  PolynomialRedosSink() { regex_match(reg.getRegex(), this.asExpr()) }
+
+  RegExpTerm getRegExp() { result = reg }
+}
+
+class PolynomialRedosConfig extends DataFlow::Configuration {
```

Comment:
Should this be `TaintTracking`? Most `RemoteFlowSource`s aren't necessarily strings, so without some additional taint steps then this might not catch much.